src/PowerShellEditorServices/Services/PowerShell/Host/InternalHost.cs

```diff
@@ -0,0 +1,778 @@
+        {
+            // TODO: Handle exit code if needed
+            SetExit();
+        }
+
+        public async Task StartAsync(HostStartOptions startOptions, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Host starting");
+            if (Interlocked.Exchange(ref _isRunning, 1) != 0)
+            {
+                _logger.LogDebug("Host start requested after already started");
+                return;
+            }
+
+            _pipelineThread.Start();
+
+            if (startOptions.LoadProfiles)
+            {
+                await ExecuteDelegateAsync(
+                   "LoadProfiles",
+                   new PowerShellExecutionOptions { MustRunInForeground = true },
+                   cancellationToken,
+                    (pwsh, delegateCancellation) =>
+                    {
+                        pwsh.LoadProfiles(_hostInfo.ProfilePaths);
+                    }).ConfigureAwait(false);
+
+                _logger.LogInformation("Profiles loaded");
+            }
+
+            if (startOptions.InitialWorkingDirectory is not null)
+            {
+                await SetInitialWorkingDirectoryAsync(startOptions.InitialWorkingDirectory, CancellationToken.None).ConfigureAwait(false);
+            }
+        }
+
+        public void SetExit()
+        {
+            if (_psFrameStack.Count <= 1)
+            {
+                return;
+            }
+
+            _shouldExit = true;
+        }
+
+        public Task<T> InvokeTaskOnPipelineThreadAsync<T>(
+            SynchronousTask<T> task)
+        {
+            if (task.ExecutionOptions.InterruptCurrentForeground)
+            {
+                // When a task must displace the current foreground command,
+                // we must:
+                //  - block the consumer thread from mutating the queue
+                //  - cancel any running task on the consumer thread
+                //  - place our task on the front of the queue
+                //  - unblock the consumer thread
+                using (_taskQueue.BlockConsumers())
+                {
+                    CancelCurrentTask();
+                    _taskQueue.Prepend(task);
+                }
+
+                return task.Task;
+            }
+
+            switch (task.ExecutionOptions.Priority)
+            {
+                case ExecutionPriority.Next:
+                    _taskQueue.Prepend(task);
+                    break;
+
+                case ExecutionPriority.Normal:
+                    _taskQueue.Append(task);
+                    break;
+            }
+
+            return task.Task;
+        }
+
+        public void CancelCurrentTask()
+        {
+            _cancellationContext.CancelCurrentTask();
+        }
+
+        public Task<TResult> ExecuteDelegateAsync<TResult>(
+            string representation,
+            ExecutionOptions executionOptions,
+            CancellationToken cancellationToken,
```

Comment:
Should move the cancellation token to the last argument here, even though it's ugly to have the delegate in the middle of the parameters